File: scnserver/models/channel.go
package models

type Channel struct {
	ChannelID         ChannelID `db:"channel_id"         json:"channel_id"`
	OwnerUserID       UserID    `db:"owner_user_id"      json:"owner_user_id"`
	InternalName      string    `db:"internal_name"      json:"internal_name"`                               // = InternalName, used for sending, normalized, cannot be changed
	DisplayName       string    `db:"display_name"       json:"display_name"`                                // = DisplayName, used for display purposes, can be changed, initially equals InternalName
	DescriptionName   *string   `db:"description_name"   json:"description_name"`                            // = DescriptionName, (optional), longer description text, initally nil
	SubscribeKey      string    `db:"subscribe_key"      json:"subscribe_key"      jsonfilter:"INCLUDE_KEY"` // can be nil, depending on endpoint
	TimestampCreated  SCNTime   `db:"timestamp_created"  json:"timestamp_created"`
	TimestampLastSent *SCNTime  `db:"timestamp_lastsent" json:"timestamp_lastsent"`
	MessagesSent      int       `db:"messages_sent"      json:"messages_sent"`
}

type ChannelWithSubscription struct {
	Channel
	Subscription *Subscription `db:"sub" json:"subscription"`
}

type ChannelPreview struct {
	ChannelID       ChannelID `json:"channel_id"`
	OwnerUserID     UserID    `json:"owner_user_id"`
	InternalName    string    `json:"internal_name"`
	DisplayName     string    `json:"display_name"`
	DescriptionName *string   `json:"description_name"`
}

func (c Channel) WithSubscription(sub *Subscription) ChannelWithSubscription {
	return ChannelWithSubscription{
		Channel:      c,
		Subscription: sub,
	}
}

func (c Channel) Preview() ChannelPreview {
	return ChannelPreview{
		ChannelID:       c.ChannelID,
		OwnerUserID:     c.OwnerUserID,
		InternalName:    c.InternalName,
		DisplayName:     c.DisplayName,
		DescriptionName: c.DescriptionName,
	}
}

File: flutter/lib/models/channel.dart
import 'package:hive_flutter/hive_flutter.dart';
import 'package:simplecloudnotifier/models/subscription.dart';
import 'package:simplecloudnotifier/state/interfaces.dart';

part 'channel.g.dart';

@HiveType(typeId: 104)
class Channel extends HiveObject implements FieldDebuggable {
  @HiveField(0)
  final String channelID;

  @HiveField(10)
  final String ownerUserID;
  @HiveField(11)
  final String internalName; // = InternalName, used for sending, normalized, cannot be changed
  @HiveField(12)
  final String displayName; // = DisplayName, used for display purposes, can be changed, initially equals InternalName
  @HiveField(13)
  final String? descriptionName; // = DescriptionName, (optional), longer description text, initally nil
  @HiveField(14)
  final String? subscribeKey;
  @HiveField(15)
  final String timestampCreated;
  @HiveField(16)
  final String? timestampLastSent;
  @HiveField(17)
  final int messagesSent;

  Channel({
    required this.channelID,
    required this.ownerUserID,
    required this.internalName,
    required this.displayName,
    required this.descriptionName,
    required this.subscribeKey,
    required this.timestampCreated,
    required this.timestampLastSent,
    required this.messagesSent,
  });

  factory Channel.fromJson(Map<String, dynamic> json) {
    return Channel(
      channelID: json['channel_id'] as String,
      ownerUserID: json['owner_user_id'] as String,
      internalName: json['internal_name'] as String,
      displayName: json['display_name'] as String,
      descriptionName: json['description_name'] as String?,
      subscribeKey: json['subscribe_key'] as String?,
      timestampCreated: json['timestamp_created'] as String,
      timestampLastSent: json['timestamp_lastsent'] as String?,
      messagesSent: json['messages_sent'] as int,
    );
  }

  @override
  String toString() {
    return 'Channel[${this.channelID}]';
  }

  List<(String, String)> debugFieldList() {
    return [
      ('channelID', this.channelID),
      ('ownerUserID', this.ownerUserID),
      ('internalName', this.internalName),
      ('displayName', this.displayName),
      ('descriptionName', this.descriptionName ?? ''),
      ('subscribeKey', this.subscribeKey ?? ''),
      ('timestampCreated', this.timestampCreated),
      ('timestampLastSent', this.timestampLastSent ?? ''),
      ('messagesSent', '${this.messagesSent}'),
    ];
  }

  ChannelPreview toPreview() {
    return ChannelPreview(
      channelID: this.channelID,
      ownerUserID: this.ownerUserID,
      internalName: this.internalName,
      displayName: this.displayName,
      descriptionName: this.descriptionName,
    );
  }
}

class ChannelWithSubscription {
  final Channel channel;
  final Subscription? subscription;

  ChannelWithSubscription({
    required this.channel,
    required this.subscription,
  });

  factory ChannelWithSubscription.fromJson(Map<String, dynamic> json) {
    return ChannelWithSubscription(
      channel: Channel.fromJson(json),
      subscription: json['subscription'] == null ? null : Subscription.fromJson(json['subscription'] as Map<String, dynamic>),
    );
  }

  static List<ChannelWithSubscription> fromJsonArray(List<dynamic> jsonArr) {
    return jsonArr.map<ChannelWithSubscription>((e) => ChannelWithSubscription.fromJson(e as Map<String, dynamic>)).toList();
  }
}

class ChannelPreview {
  final String channelID;
  final String ownerUserID;
  final String internalName;
  final String displayName;
  final String? descriptionName;

  const ChannelPreview({
    required this.channelID,
    required this.ownerUserID,
    required this.internalName,
    required this.displayName,
    required this.descriptionName,
  });

  factory ChannelPreview.fromJson(Map<String, dynamic> json) {
    return ChannelPreview(
      channelID: json['channel_id'] as String,
      ownerUserID: json['owner_user_id'] as String,
      internalName: json['internal_name'] as String,
      displayName: json['display_name'] as String,
      descriptionName: json['description_name'] as String?,
    );
  }
}
